src/Plugin/Validators/ValidHeaders.php

```diff
@@ -86,18 +128,151 @@ public function validateRow($headers) {
+   *
+   * @param array $validation_status
+   *   An associative array that was returned by the ValidHeaders validator in
+   *   the event of failed validation. It contains the following keys:
+   *   - 'case': a developer-focused string describing the case checked.
+   *   - 'valid': FALSE to indicate that validation failed.
+   *   - 'failedItems': an array of items that failed, either:
+   *     - 'headers': A string indicating the header row is empty.
+   *     - an array of column headers that was in the input file.
+   *       @see validateRow()
+   * @param array $metadata
+   *   An array of additional metadata (or contextual information) needed by the
+   *   process method. Here, the following keys are expected:
+   *   - 'column_headers': This contains an array of headers defined in the
+   *     importer. The index in this array MUST match the position
+   *     (starting with 0) of the column in the input file.
+   *     Eg: 'column_headers' => [
+   *           '2' => 'Header 2', // Header of column #3
+   *           '4' => 'Header 4', // Header of column #5
+   *         ];.
+   * @param array $tokens
+   *   [OPTIONAL] An array of values to use for token replacement.
+   *   @see ValidHeaders::$mapping
+   *   The following token keys will substitute the entire existing case message
+   *   to the user with the value of that token.
+   *   - 'case-empty-headers': the message when there is no header row.
+   *   - 'case-mismatch-values': the message when headers are not expected.
+   *   - 'case-mismatch-count': the message when headers count is not expected.
+   *   - 'num-expected-columns': number of expected column headers.
+   *
+   * @return array
+   *   A render array of type unordered list which is used to display feedback
+   *   to the user about the case that failed and the failed items from the
+   *   input file. This unordered list will include a table with a row of the
+   *   expected headers followed by a row of the provided headers.
+   *
+   * @throws \Exception
+   *   - If the validation_status parameter was not formatted properly.
+   *   - If the case string returned by the validator implied validation passed.
+   *   - If the case string returned by the validator is not recognized.
+   */
+  public static function processValidHeadersFailures(array $validation_status, array $metadata, array $tokens = []) {
+
+    // Check the format of the validation_status parameter.
+    ImportValidationHelper::checkValidationStatusArray($validation_status, 'ValidHeaders');
+
+    $default_tokens = array_column(self::$mapping, 'default-msg', 'token');
+    // Combine our provided and our default token arrays. Because array_merge
+    // will overwrite values in the first array with values from the second
+    // array for the same keys, we provide our default tokens first.
+    $combined_tokens = array_merge($default_tokens, $tokens);
+
+    $combined_tokens['num-expected-columns'] = count($metadata['column_headers']);
```

Comment:
$metadata needs to be checked for this index first, and throw an exception if it doesn't exist.